compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/EffectiveTypePopulator.java

```diff
@@ -0,0 +1,482 @@
+
+                for (BType memberType : originalMemberList) {
+                    if (types.isInherentlyImmutableType(memberType)) {
+                        bUnionType.add(memberType);
+                        continue;
+                    }
+
+                    BType immutableMemberType = ImmutableTypeCloner.getImmutableType(loc, types, memberType, env, pkgID,
+                            env.scope.owner, symTable, anonymousModelHelper, names, new HashSet<>());
+                    bUnionType.add(immutableMemberType);
+                }
+                return;
+            }
+        }
+
+        for (BType immutableMemberType : bUnionType.getMemberTypes()) {
+            updateType(immutableMemberType, loc, pkgID, typeNode, env);
+        }
+        bUnionType.mutableType = null;
+    }
+
+    @Override
+    public void visit(BIntersectionType bIntersectionType) {
+        updateType(bIntersectionType.effectiveType, loc, pkgID, typeNode, env);
+    }
+
+    @Override
+    public void visit(BXMLType bXMLType) {
+        BXMLType origXMLType = bXMLType.mutableType;
+        if (origXMLType != null) {
+            if (bXMLType.constraint == symTable.semanticError) {
+                bXMLType.constraint = ImmutableTypeCloner.getImmutableType(loc, types, origXMLType.constraint, env,
+                        pkgID, env.scope.owner, symTable, anonymousModelHelper, names, new HashSet<>());
+            }
+            updateType(bXMLType.constraint, loc, pkgID, typeNode, env);
+            bXMLType.mutableType = null;
+        }
+    }
+
+    @Override
+    public void visit(BTableType bTableType) {
+        BTableType origTableType = bTableType.mutableType;
+        if (origTableType != null) {
+            if (bTableType.constraint.tag == TypeTags.NEVER ||
+                    bTableType.constraint == symTable.semanticError) {
+                bTableType.constraint = ImmutableTypeCloner.getImmutableType(loc, types, origTableType.constraint, env,
+                        pkgID, env.scope.owner, symTable, anonymousModelHelper, names, new HashSet<>());
+            }
+            updateType(bTableType.constraint, loc, pkgID, typeNode, env);
+
+            if (origTableType.keyTypeConstraint != null) {
+                bTableType.keyTypeConstraint = ImmutableTypeCloner.getImmutableType(loc, types,
+                        origTableType.keyTypeConstraint, env, pkgID, env.scope.owner, symTable,
+                        anonymousModelHelper, names, new HashSet<>());
+                updateType(bTableType.keyTypeConstraint, loc, pkgID, typeNode, env);
+            }
+            bTableType.mutableType = null;
+        }
+    }
+
+    @Override
+    public void visit(BRecordType bRecordType) {
+        BRecordType origRecordType = bRecordType.mutableType;
+        BTypeSymbol structureSymbol = bRecordType.tsymbol;
+        if (origRecordType != null) {
+            if (origRecordType.fields.size() != bRecordType.fields.size()) {
```

Comment:
Why do we need to do this? can we just take an immutable type clone of the `origRecordType` and populate the `origRecordType`?

Reply:
We can't do that since we using cache in the immutable cloner https://github.com/ballerina-platform/ballerina-lang/blob/3a4e6354f1030b19aa6b4559b80aa0660bc41def/compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/util/ImmutableTypeCloner.java#L197 
when we taking the immutable clone it will return the already created immutable clone which is incorrect.

Reply:
Let us overload `getImmutableType` to not look at the `immutableTypeMaps` in this case. @MaryamZi I think we need to rethink keeping the `immutableTypeMaps` also

Reply:
If we remove the immutable type map we will define a new object/record type (+ type definition) wherever there is a readonly intersection `readonly & T`, which will result in new value classes at runtime which is overhead. This is basically for effective types. 

Why do we need to remove it? I'm not sure if overloading is correct since it can lead to inconsistencies (different types being referred to).

Reply:
What do u mean by `different types being referred to`?

Reply:
If someone calls getImmutableType/gets the effective type before it is completely resolved, and we set a new immutable/effective type once it can be completely resolved, the first type will be incorrect.

Reply:
What I'm suggesting is we don't touch the `immutableTypeMaps` in the overridden method and calculate the correct effective type. Then populate the incorrect effective type with the correct one. At that time `immutableTypeMaps` will also be updated because it contains the same reference

Reply:
You mean without looking fieldwise? Could work, we'll be creating an extra type I guess, but if it reduces special casing and is temporary should be OK. I believe we also discussed this once @prakanth97.

> What I'm suggesting is we don't touch the `immutableTypeMaps` in the overridden method

Populating the type in `immutableTypeMaps` will also have to be done here though? At the end maybe.

Reply:
Yes, it reduces the duplicate work in this method. `Populating the type in immutableTypeMaps will also have to be done here though? At the end maybe.` that will be automatically updated when we update incorrect effective type since its the same reference

Reply:
Will address this when I add the generic intersection support

Reply:
> when we update incorrect effective type since its the same reference

This is what I meant. That has to be done here? We have to populate the fields from this temporary type?

Reply:
Yes